Sources/Sentry/SentryProfiler.mm

```diff
@@ -183,63 +181,109 @@
+            SENTRY_LOG_DEBUG(@"Frame render finished after transaction finished, won't record.");
+            return;
+        }
+        const auto relativeFrameRenderStart
+            = getDurationNs(transaction.startSystemTime, frameRenderStart);
+        const auto relativeFrameRenderEnd
+            = getDurationNs(transaction.startSystemTime, frameRenderEnd);
+
+        // this probably won't happen, but doesn't hurt to have one last defensive check before
+        // calling getDurationNs
+        if (!orderedChronologically(relativeFrameRenderStart, relativeFrameRenderEnd)) {
+            SENTRY_LOG_WARN(
+                @"Computed relative start and end timestamps are not chronologically ordered.");
             return;
         }
-        const auto frameRenderStartRelativeToProfileStartNs
-            = getDurationNs(profileStart, frameRenderStart);
         const auto frameRenderDurationNs
-            = frameRenderEndRelativeToProfileStart - frameRenderStartRelativeToProfileStartNs;
+            = getDurationNs(relativeFrameRenderStart, relativeFrameRenderEnd);
+
         [relativeFrameInfo addObject:@{
-            @"elapsed_since_start_ns" : @(frameRenderStartRelativeToProfileStartNs),
+            @"elapsed_since_start_ns" :
+                [NSString stringWithFormat:@"%llu", relativeFrameRenderStart],
             @"value" : @(frameRenderDurationNs),
         }];
     }];
     return relativeFrameInfo;
 }
 
 /**
- * Convert the data structure that records timestamps for CPU frame rate info from
+ * Convert the data structure that records timestamps for GPU frame rate info from
  * SentryFramesTracker to the structure expected for profiling metrics.
  */
 NSArray<NSDictionary *> *
-processFrameRates(SentryFrameInfoTimeSeries *frameRates, uint64_t start)
+processFrameRates(SentryFrameInfoTimeSeries *frameRates, SentryTransaction *transaction)
 {
     auto relativeFrameRates = [NSMutableArray array];
     [frameRates enumerateObjectsUsingBlock:^(
         NSDictionary<NSString *, NSNumber *> *_Nonnull obj, NSUInteger idx, BOOL *_Nonnull stop) {
         const auto timestamp = obj[@"timestamp"].unsignedLongLongValue;
         const auto refreshRate = obj[@"frame_rate"];
-        uint64_t relativeTimestamp = 0;
-        if (timestamp >= start) {
-            relativeTimestamp = getDurationNs(start, timestamp);
+
+        if (!orderedChronologically(transaction.startSystemTime, timestamp)) {
+            return;
         }
-        [relativeFrameRates addObject:@{
-            @"elapsed_since_start_ns" : @(relativeTimestamp),
+        if (orderedChronologically(transaction.endSystemTime, timestamp)) {
+            return;
+        }
+
+        const auto relativeTimestamp = getDurationNs(transaction.startSystemTime, timestamp);
+
+        [relativeFrameRates addObject:@ {
+            @"elapsed_since_start_ns" : [NSString stringWithFormat:@"%llu", relativeTimestamp],
```

Comment:
Interesting tidbit here, we should always be sending a string for `elapsed_since_start_ns` since it's an unsigned 64 bit integer which JSON technically doesn't support! Somehow this crept in as an `NSNumber` previously... probably isn't ever failing in practice, but it's nicer to be consistent.

Reply:
`l`: Maybe it makes sense to create a function for `[NSString stringWithFormat:@"%llu", timestamp]` so it's easier to change.

Reply:
Done! https://github.com/getsentry/sentry-cocoa/pull/2729/commits/dffd985bf73f25c7f00d4d726aa4ae00f9b2c257